File: backend/auth/src/routers/__tests__/signup.router.test.js
const request = require("supertest");
const express = require("express");
const AuthService = require("../../services/auth.service");
const rootRooter = require("../../routers");
const { db } = require("decorebator-common");

describe("Signup endpoint tests", () => {
  let app;

  beforeAll(async () => {
    await db.connect();
    app = express();
    app.use("/", rootRooter);
  });

  afterAll(async () => {
    await db.disconnect();
  });

  beforeEach(async () => {
    await AuthService.removeAccount("signup.test@gmail.com");
    await AuthService.removeAccount("signup.test2@gmail.com");
  });

  it("Should be able to register a new user", done => {
    request(app)
      .post("/signup")
      .set("content-type", "application/json")
      .send({ login: "signup.test@gmail.com", name: "Lucas Simão", password: "123456789", country: "BR" })
      .expect(200, done);
  });

  it("should deny the registering of duplicate users", async done => {
    const user = {
      login: "signup.test2@gmail.com",
      name: "Lucas Simão",
      password: "123456789",
      country: "BR"
    };

    await request(app)
      .post("/signup")
      .set("content-type", "application/json")
      .send(user)
      .expect(200);

    request(app)
      .post("/signup")
      .set("content-type", "application/json")
      .send(user)
      .expect(400, "User already exists", done);
  });

  it("should deny registering with a invalid email", done => {
    const user = {
      login: "my login",
      name: "Lucas Simão",
      password: "123456789",
      country: "BR"
    };

    request(app)
      .post("/signup")
      .set("content-type", "application/json")
      .send(user)
      .expect(400, "A valid email must be used as your account login", done);
  });

  it("should forbid unknown countries", done => {
    const user = {
      login: "signup.test@gmail.com",
      name: "Lucas Simão",
      password: "123456789",
      country: "INEXISTING"
    };

    request(app)
      .post("/signup")
      .set("content-type", "application/json")
      .send(user)
      .expect(400, "Invalid country", done);
  });
});

File: backend/integration-tests/src/test/java/com/decorebator/integration/auth/SignUpRouterTest.java
package com.decorebator.integration.auth;

import static io.restassured.RestAssured.given;
import static org.hamcrest.Matchers.equalTo;

import com.decorebator.beans.UserRegistration;
import com.decorebator.integration.EnvironmentRule;

import org.junit.BeforeClass;
import org.junit.ClassRule;
import org.junit.Test;

import io.restassured.RestAssured;
import io.restassured.http.ContentType;

/**
 * 
 * Integration tests for /signup microservice
 */
public class SignUpRouterTest {

    @ClassRule
    public static EnvironmentRule environmentRule = new EnvironmentRule(true);

    @BeforeClass
    public static void setup() {
        RestAssured.enableLoggingOfRequestAndResponseIfValidationFails();
        RestAssured.baseURI = environmentRule.getAuthHostAndPort();
    }

    @Test
    public void shouldBeAbleToRegisterANewUser() {
        var userRegistration  = new UserRegistration("lsimaocosta@gmail.com", "Lucas Simão","123456789","BR" );

        given()
            .contentType(ContentType.JSON)
            .body(userRegistration)
        .when()
            .post("/signup")
        .then()
            .statusCode(200);
    }

    @Test
    public void shouldDenyRegistrationWithSameEmail() {
        var userRegistration  = new UserRegistration("sigup.test@gmail.com", "Lucas Simão","123456789","BR" );
        var userRegistration2  = new UserRegistration("sigup.test@gmail.com", "any name","sssss","BR" );

        given()
            .contentType(ContentType.JSON)
            .body(userRegistration)
        .when()
            .post("/signup")
        .then()
            .statusCode(200);


        given()
            .contentType(ContentType.JSON)
            .body(userRegistration2)
        .when()
            .post("/signup")
        .then()
            .statusCode(400)
            .body(equalTo("User already exists"));
    }

    @Test
    public void shouldDenyRegisteringWithInvalidEmail() {
        var userRegistration  = new UserRegistration("invalid email", "Lucas Simão","123456789","BR" );

        given()
            .contentType(ContentType.JSON)
            .body(userRegistration)
        .when()
            .post("/signup")
        .then()
            .statusCode(400)
            .body(equalTo("A valid email must be used as your account login"));    
    }


    @Test
    public void shouldDenyRegisteringWithInvalidCountryCode() {
        var userRegistration  = new UserRegistration("sigup.test@gmail.com", "any name","any pawd","wrong country code" );

        given()
            .contentType(ContentType.JSON)
            .body(userRegistration)
        .when()
            .post("/signup")
        .then()
            .statusCode(400)
            .body(equalTo("Invalid country"));    
    }    

}